modules/base/src/main/scala/io/janstenpickle/trace4cats/base/context/Inject.scala

```diff
@@ -0,0 +1,12 @@
+package io.janstenpickle.trace4cats.base.context
+
+trait Inject[F[_], R, R0] { self =>
```

Comment:
What is the common sense behind the `Inject` trait? I think you mixed 2 different concepts here.

As I see it, the aim of `Inject` is to encapsulate only one function: `R0 => G[R]` where `G[_]` is something usable: typically a resource over F, or (rarely) is F itself.

So given instances 
```scala
val inject: Inject[F, R, R0] = ???
val r0: R0 = ???
val f: R => F[A] = ???
```
a call to inject:
```scala
inject.apply(r0)(f)
```
may be replaced with
```scala
makeR(r0).use(f)
```
where
```scala
def makeR: R0 => Usable[R] = ???  //this is what an `Inject` instance actually encapsulates
```
where `Usable[R]` is `Resource[F, R]` or (rarely) `F[R]` with use=flatMap.

We can even introduce a new trait:
```scala
trait Usable[F[_], A] {
  def use[B](f: A => F[B]): F[B]
}

object Usable {
  def fromEffect[F[_]: Monad, A](res: F[A]): Usable[F, A] =
    new Usable[F, A] {
      def use[B](f: A => F[B]): F[B] = res.flatMap(f)
    }

  def fromResource[F[_]: BracketThrow, A](res: Resource[F, A]): Usable[F, A] =
    new Usable[F, A] {
      def use[B](f: A => F[B]): F[B] = res.use(f)
    }
}
```

But I think it's overkill and we can just go ahead with the concrete `Resource[F, A]`, because we use it for spans only, and a span is always a resource.

I think the second part of the curried `apply`(`(f: R => F[A]): F[A]`)  is redundant here. And only the first one is meaningful. So in fact, `Inject` is just a resource provider. And the resource itself has the ability to be used with a function `f`.

So, does a simple provider function deserve a separate typeclass?
What about 
```scala
type SpanProvider[F[_], R] = R => Resource[F, Span[F]]
```

Reply:
With all said above, I think both `Inject` and `Init` are not simple typeclasses, but some overcomplicated constructions that may be divided into more simple things. Moreover, they don't belong to the `context` package or maybe even the `base` module.

The original idea behind the context package was to create typeclass hierarchy for contextual effects F[_], such that forall `A`: `F[A]` is isomorphic to the `R => Low[A]` arrow. Among them are `cats.data.Kleisli`, `zio.ZIO`, `tofu.env.Env`, etc.

Reply:
Thanks @catostrophe this all makes sense. I had my reservations about making a typeclass too, especially as, you pointed out, it behaves more as a constructor. I will try and do something more in `inject` which can be used with the entrypoint.